vulcanai2/models/dnn.py

```diff
@@ -144,12 +146,13 @@ def _create_network(self, **kwargs):
                 kwargs['pred_activation'])
 
     def _create_classification_layer(self, dim, pred_activation):
-        self.network_tail = nn.Sequential(
-                FlattenUnit(),
-                DenseUnit(
-                    in_features=dim,
-                    out_features=self._num_classes,
-                    activation=pred_activation))
+        self.network.add_module(
+            'flatten', FlattenUnit())
+        self.network.add_module(
+            'classify', DenseUnit(
+                in_features=dim,
+                out_features=self._num_classes,
+                activation=pred_activation))
```

Comment:
same: I would suggest doing this within _build_network itself

Reply:
the issue is that get_flattened_size requires a `self.network` to already exist @priyatharsan

Reply:
oh dont worry about that. At this moment lets make it optional to compute flattened size only if the network does not have a `network_tail`. We were using flattened_size for shape inferencing between `ConvNet` and `DenseNet` outside of the `BaseNetwork` container, remember? Now it is nearly not required to have that computation.

Reply:
if we don't know the flattened size then how would you set up your last DenseUnit? How else would you determine the in_features?

Reply:
Ah I see you are right! How about we make it like this:

```

    def _create_network(self, **kwargs):
        conv_hid_layers = self._config.units

        if self.input_networks is not None:
            # Create empty input tensors
            in_tensors = []
            for d in self.input_networks:
                in_tensors.append(torch.ones([1, *d.out_dim]))
            output = self._merge_input_network_outputs(in_tensors)
            self.in_dim = tuple(output.shape[1:])

        # Build Network
        self.network = self._build_conv_network(
            conv_hid_layers, kwargs['activation'])


    def _build_conv_network(self, conv_hid_layers, activation):
        """
        Build the layers of the network into a nn.Sequential object.
        Parameters
        ----------
        conv_hid_layers : ConvNetConfig.units (list of dict)
            The hidden layers specification
        activation : torch.nn.Module
            the non-linear activation to apply to each layer
        Returns
        -------
        output : torch.nn.Sequential
            the conv network as a nn.Sequential object
        """
        conv_hid_layers[0]['in_channels'] = self.in_dim[0]
        conv_layers = OrderedDict()
        for idx, conv_layer_config in enumerate(conv_hid_layers):
            conv_layer_config['activation'] = activation
            layer_name = 'conv_{}'.format(idx)
            conv_layers[layer_name] = ConvUnit(**conv_layer_config)
        conv_network = nn.Sequential(conv_layers)

        if self._num_classes:
           conv_network .add_module(
                   'flatten', FlattenUnit())
           conv_network .add_module(
                'classify', DenseUnit(
                    in_features=self.get_flattened_size(),
                    out_features=self._num_classes,
                    activation=pred_activation))

       return conv_network

Reply:
yeah that's what I tried but it will crash as soon as you run `self.get_flattened_size()`. self.network doesn't yet exist

Reply:
Yeah I did a slight modification now ⬆️

Reply:
whats more funny is you can further replace `in_features=self.get_flattened_size()` with `in_features=self._get_out_dim()[0]` 😁  and remove self.get_flattened_size

Reply:
I see. ill give it a shot!